File: golang/_0017_Letter_Combinations_of_a_Phone_Number/main.go
package main

var mp = map[byte]string{
	'2': "abc",
	'3': "def",
	'4': "ghi",
	'5': "jkl",
	'6': "mno",
	'7': "pqrs",
	'8': "tuv",
	'9': "wxyz",
}

func letterCombinations(digits string) []string {
	res, curRes := make([]string, 0), ""
	if len(digits) > 0 {
		dfs(&res, curRes, mp, digits, 0)
	}

	return res
}

func dfs(res *[]string, curRes string, mp map[byte]string, digits string, pos int) {
	if pos == len(digits) {
		*res = append(*res, curRes)
		return
	}

	for _, v := range mp[digits[pos]] {
		dfs(res, curRes+string(v), mp, digits, pos+1)
	}
}

File: python3/_0017_Letter_Combinations_of_a_Phone_Number/main.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        res = []
        self.dfs(res, "", digits, 0)

        return res

    def dfs(self, res: List[str], oneRes: str, digits: str, pos: int):
        if pos == len(digits):
            res.append(oneRes)
            return

        dt = {
            '2': "abc",
            '3': "def",
            '4': "ghi",
            '5': "jkl",
            '6': "mno",
            '7': "pqrs",
            '8': "tuv",
            '9': "wxyz",
        }

        digit = digits[pos]
        for c in dt.get(digit):
            self.dfs(res, oneRes + c, digits, pos + 1)
